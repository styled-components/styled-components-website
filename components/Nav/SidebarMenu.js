import React, { Component } from 'react'
import styled from 'styled-components'
import { I18n } from 'react-i18next'

import rem from '../../utils/rem'
import Link, { StyledLink } from '../Link'
import titleToDash from '../../utils/titleToDash'
// import { lightGrey } from '../../utils/colors'
// import { mobile } from '../../utils/media'

import { pages } from '../../pages/docs.json'

import {
  NAV_TRANSLATION,
} from '../../constants/i18n'

const MenuInner = styled.div`
  display: block;
  box-sizing: border-box;
  height: 100%;
  padding-top: ${rem(25)};
`

const Section = styled.div`
  margin-bottom: ${rem(20)};
`

const SectionTitle = styled.h4`
  display: block;
  margin: ${rem(10)} ${rem(40)};
  font-weight: normal;
`

const SubSection = styled.h5`
  display: block;
  margin: ${rem(10)} ${rem(40)} ${rem(10)} ${rem(55)};
  font-size: 0.9rem;
  font-weight: normal;
`

class Folder extends Component {
  state = {
    isOpen: false,
  }

  toggleSubSections = () => {
    this.setState({ isOpen: !this.state.isOpen })
  }

  componentWillMount() {
    this.setState({ isOpen: this.props.isOpenDefault })
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ isOpen: nextProps.isOpenDefault })
  }

  render() {
    // eslint-disable-next-line
    const { children, isOpenDefault, ...props } = this.props
    const { isOpen } = this.state

    return typeof children === 'function'
      ? children({ rootProps: props, toggleSubSections: this.toggleSubSections, isOpen })
      : null
  }
}

const SidebarMenu = ({ onRouteChange }) => (
  <I18n
    ns={NAV_TRANSLATION}
  >
    {translate => (
      <MenuInner>
        {
          pages.map(({ title, pathname, sections }) => (
              <Folder
                key={title}
                isOpenDefault={
                  typeof window !== 'undefined' &&
                    (window.location.pathname === `/docs/${pathname}`)
                }
              >
                {({
                  rootProps,
                  toggleSubSections,
                  isOpen,
                }) => (
                  <Section {...rootProps} onClick={onRouteChange}>
                    <SectionTitle onClick={toggleSubSections}>
                      <Link href={`/docs/${pathname}`}>
                        {translate(title)}
                      </Link>
                    </SectionTitle>

                    {
                      isOpen && sections.map(({ title }) => (
                        <SubSection key={title}>
                          <StyledLink href={`/docs/${pathname}#${titleToDash(translate(title))}`}>
                            {translate(title)}
                          </StyledLink>
                        </SubSection>
                      ))
                    }
                    </Section>
                )}
              </Folder>
          ))
        }
      </MenuInner>
    )}
  </I18n>
)

export default (SidebarMenu)
