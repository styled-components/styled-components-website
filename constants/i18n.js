const DEFAULT_TRANSLATION = 'translations'
const NAV_TRANSLATION = 'nav'
const HOME_TRANSLATION = 'home'
const DOCS_TRANSLATION = 'docs'

const TRANSLATIONS = [
  DEFAULT_TRANSLATION,
  NAV_TRANSLATION,
  HOME_TRANSLATION,
  DOCS_TRANSLATION,
]

const DEFAULT_LANGUAGE = 'en'

const LANGUAGES = [
  DEFAULT_LANGUAGE,
  'es',
]

module.exports = {
  DEFAULT_TRANSLATION,
  NAV_TRANSLATION,
  TRANSLATIONS,
  HOME_TRANSLATION,
  DOCS_TRANSLATION,
  LANGUAGES,
  DEFAULT_LANGUAGE,
}
