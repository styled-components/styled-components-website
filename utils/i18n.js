import i18n from 'i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import { reactI18nextModule } from 'react-i18next'

import translations from '../locales'

i18n
  .use(LanguageDetector)
  .use(reactI18nextModule)
  .init({
    fallbackLng: 'en',

    // have a common namespace used around the full app
    ns: ['translations', 'homepage'],
    defaultNS: 'translations',

    debug: process.env.NODE_ENV !== 'production',

    interpolation: {
      escapeValue: false, // not needed for react!!
    },

    react: {
      wait: true
    },

    resources: { ...translations }
  })

export default i18n
