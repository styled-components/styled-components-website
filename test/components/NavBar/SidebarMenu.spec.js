import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import 'jest-styled-components'
import { I18nextProvider } from 'react-i18next'
import SidebarMenu from '../../../components/Nav/SidebarMenu'

import i18n from '../../helpers/i18n'

const wrapper = (language, props) => mount(
  <I18nextProvider i18n={i18n} initialLanguage={language}>
    <SidebarMenu />
  </I18nextProvider>
)

test('SidebarMenu renders correctly', () => {
  const component = wrapper('en')
  expect(component.find(SidebarMenu)).toMatchSnapshot()
})
