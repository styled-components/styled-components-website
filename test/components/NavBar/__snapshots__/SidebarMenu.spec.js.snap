// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarMenu renders correctly 1`] = `
.c3 {
  display: inline-block;
  text-decoration: none;
  color: inherit;
  cursor: pointer;
  padding: 0.1111111111111111rem 0.4444444444444444rem;
  margin: -0.1111111111111111rem -0.4444444444444444rem;
}

.c0 {
  display: block;
  box-sizing: border-box;
  height: 100%;
  padding-top: 1.3888888888888888rem;
}

.c1 {
  margin-bottom: 1.1111111111111112rem;
}

.c2 {
  display: block;
  margin: 0.5555555555555556rem 2.2222222222222223rem;
  font-weight: normal;
}

@media (min-width:62.5em) {
  .c3 {
    border-radius: 0.16666666666666666rem;
  }

  .c3:hover {
    background: rgba(20,20,20,0.1);
  }
}

<SidebarMenu>
  <I18n
    ns="nav"
  >
    <styled.div>
      <div
        className="c0"
      >
        <Folder
          isOpenDefault={false}
        >
          <styled.div>
            <div
              className="c1"
            >
              <styled.h4
                onClick={[Function]}
              >
                <h4
                  className="c2"
                  onClick={[Function]}
                >
                  <Link
                    href="/docs/basics"
                  >
                    <Link
                      href="/docs/basics"
                    >
                      <styled.a
                        href="/docs/basics"
                        onClick={[Function]}
                      >
                        <a
                          className="c3"
                          href="/docs/basics"
                          onClick={[Function]}
                        >
                          Basics
                        </a>
                      </styled.a>
                    </Link>
                  </Link>
                </h4>
              </styled.h4>
            </div>
          </styled.div>
        </Folder>
        <Folder
          isOpenDefault={false}
        >
          <styled.div>
            <div
              className="c1"
            >
              <styled.h4
                onClick={[Function]}
              >
                <h4
                  className="c2"
                  onClick={[Function]}
                >
                  <Link
                    href="/docs/advanced"
                  >
                    <Link
                      href="/docs/advanced"
                    >
                      <styled.a
                        href="/docs/advanced"
                        onClick={[Function]}
                      >
                        <a
                          className="c3"
                          href="/docs/advanced"
                          onClick={[Function]}
                        >
                          Advanced
                        </a>
                      </styled.a>
                    </Link>
                  </Link>
                </h4>
              </styled.h4>
            </div>
          </styled.div>
        </Folder>
        <Folder
          isOpenDefault={false}
        >
          <styled.div>
            <div
              className="c1"
            >
              <styled.h4
                onClick={[Function]}
              >
                <h4
                  className="c2"
                  onClick={[Function]}
                >
                  <Link
                    href="/docs/api"
                  >
                    <Link
                      href="/docs/api"
                    >
                      <styled.a
                        href="/docs/api"
                        onClick={[Function]}
                      >
                        <a
                          className="c3"
                          href="/docs/api"
                          onClick={[Function]}
                        >
                          API Reference
                        </a>
                      </styled.a>
                    </Link>
                  </Link>
                </h4>
              </styled.h4>
            </div>
          </styled.div>
        </Folder>
        <Folder
          isOpenDefault={false}
        >
          <styled.div>
            <div
              className="c1"
            >
              <styled.h4
                onClick={[Function]}
              >
                <h4
                  className="c2"
                  onClick={[Function]}
                >
                  <Link
                    href="/docs/tooling"
                  >
                    <Link
                      href="/docs/tooling"
                    >
                      <styled.a
                        href="/docs/tooling"
                        onClick={[Function]}
                      >
                        <a
                          className="c3"
                          href="/docs/tooling"
                          onClick={[Function]}
                        >
                          Tooling
                        </a>
                      </styled.a>
                    </Link>
                  </Link>
                </h4>
              </styled.h4>
            </div>
          </styled.div>
        </Folder>
        <Folder
          isOpenDefault={false}
        >
          <styled.div>
            <div
              className="c1"
            >
              <styled.h4
                onClick={[Function]}
              >
                <h4
                  className="c2"
                  onClick={[Function]}
                >
                  <Link
                    href="/docs/faqs"
                  >
                    <Link
                      href="/docs/faqs"
                    >
                      <styled.a
                        href="/docs/faqs"
                        onClick={[Function]}
                      >
                        <a
                          className="c3"
                          href="/docs/faqs"
                          onClick={[Function]}
                        >
                          FAQs
                        </a>
                      </styled.a>
                    </Link>
                  </Link>
                </h4>
              </styled.h4>
            </div>
          </styled.div>
        </Folder>
      </div>
    </styled.div>
  </I18n>
</SidebarMenu>
`;
